test/indexer/test_hca_indexer.py

```diff
@@ -1256,6 +1256,97 @@ def test_accessions_fields(self):
             self.assertEqual(['E-AAAA-00'], project['array_express_accessions'])
             self.assertEqual(['PRJNA000000'], project['insdc_study_accessions'])
 
+    def test_cell_counts(self):
+        """
+        Verify the cell counts found in project, cell_suspension, and file entities
+        """
+        # Bundles from the canned staging area, both for project 90bf705c
+        # https://github.com/HumanCellAtlas/schema-test-data/
+        bundle_fqid = self.bundle_fqid(uuid='4da04038-adab-59a9-b6c4-3a61242cc972',
+                                       version='2021-01-01T00:00:00.000000Z')
+        self._index_canned_bundle(bundle_fqid)
+        bundle_fqid = self.bundle_fqid(uuid='d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7',
+                                       version='2021-01-01T00:00:00.000000Z')
+        self._index_canned_bundle(bundle_fqid)
+        hits = self._get_all_hits()
+        keys = [
+            ('projects', 'estimated_cell_count'),
+            ('cell_suspensions', 'total_estimated_cells'),
+            ('files', 'matrix_cell_count')
+        ]
+        actual_aggregate_counts = {}
+        actual_contribution_counts = {}
+        for hit in sorted(hits, key=lambda d: d['_id']):
+            contents = hit['_source']['contents']
+            entity_type, aggregate = self._parse_index_name(hit)
+            if aggregate:
+                if entity_type == 'projects':
+                    for entity_type, field_name in keys:
+                        entity = one(contents[entity_type])
+                        actual_aggregate_counts[entity_type, field_name] = entity[field_name]
+            else:
+                cell_counts = {}
+                for entity_type, field_name in keys:
+                    for entity in contents[entity_type]:
+                        entity_id = entity['document_id']
+                        cell_counts[entity_id] = entity[field_name]
+                contribution_id = hit['_source']['entity_id']
+                actual_contribution_counts[contribution_id] = cell_counts
+        expected_aggregate_counts = {
+            ('projects', 'estimated_cell_count'): 10000,
+            ('cell_suspensions', 'total_estimated_cells'): 40000,
+            ('files', 'matrix_cell_count'): 17100
+        }
+        self.assertEqual(expected_aggregate_counts, actual_aggregate_counts)
+        expected_contribution_counts = {
```

Comment:
This is too verbose. I'm certain there is a way to express this assertion more generically without having conditional or looped asserts. The strategy I've recommended in the past is to accumulate a fingerprint by looping over the results and then assert the actual fingerprint with an expected one.

Reply:
This is still a bit obscure. Can't we key on inner and outer entity type instead of document id?